components/base/kinematicbase/kinematics.go

```diff
@@ -102,56 +107,106 @@ func (ddk *differentialDriveKinematics) GoToInputs(ctx context.Context, desired
+			if poseChanged(prevDistErr, prevHeadingErr, distErr, headingErr) {
+				poseChange <- true
+			}
 
-		// get to the x, y location first - note that from the base's perspective +y is forward
-		desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
-		commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
-		if err != nil {
-			return err
-		}
+			col, err := validRegion.CollidesWith(spatialmath.NewPoint(r3.Vector{X: current[0].Value, Y: current[1].Value}, ""))
+			if err != nil {
+				movementErr <- err
+				return
+			}
+			if !col {
+				movementErr <- errors.New("base has deviated too far from path")
+				return
+			}
+
+			// get to the x, y location first - note that from the base's perspective +y is forward
+			desiredHeading := math.Atan2(current[1].Value-desired[1].Value, current[0].Value-desired[0].Value)
+			prevDistErr = distErr
+			prevHeadingErr = headingErr
+			distErr, headingErr, err = ddk.errorState(current, []referenceframe.Input{desired[0], desired[1], {desiredHeading}})
+			if err != nil {
+				movementErr <- err
+				return
+			}
+			commanded, err := ddk.issueCommand(ctx, distErr, headingErr)
+			if err != nil {
+				movementErr <- err
+				return
+			}
 
-		if !commanded {
-			// no command to move to the x, y location was issued, correct the heading and then exit
-			if commanded, err := ddk.issueCommand(ctx, current, []referenceframe.Input{current[0], current[1], desired[2]}); err == nil {
-				if !commanded {
-					return nil
+			if !commanded {
+				// no command to move to the x, y location was issued, correct the heading and then exit
+				distErr, headingErr, err = ddk.errorState(current, []referenceframe.Input{current[0], current[1], desired[2]})
+				if err != nil {
+					movementErr <- err
+					return
 				}
-			} else {
-				return err
+				if commanded, err := ddk.issueCommand(ctx, distErr, headingErr); err == nil {
+					if !commanded {
+						movementErr <- nil
+						return
+					}
+				} else {
+					movementErr <- err
+					return
+				}
+			}
+
+			current, err = ddk.CurrentInputs(ctx)
+			if err != nil {
+				movementErr <- err
+				return
 			}
 		}
+		movementErr <- contextErr
+	})
+
+	utils.PanicCapturingGo(func() {
+		for {
+			select {
+			case <-poseChange:
+				// restart timer if a position change occurs before the timer expires
+				// have to stop timer and empty channel if there's anything in it before resetting
+				if !t.Stop() {
+					<-t.C
+				}
+				t.Reset(timeout)
 
-		current, err = ddk.CurrentInputs(ctx)
-		if err != nil {
-			return err
+			case <-t.C: // timer expired before a position change occurred
+				movementErr <- errors.New("movement timeout")
+			}
 		}
-	}
-	return err
+	})
+	goToInputsErr := <-movementErr // waits on error value to be returned from either timer or movement code
+	return goToInputsErr
```

Comment:
If there's a non-nil error, we would presumably want to stop the base, right?

Reply:
I believe this will be checked in MoveOnMap and the movement This is how the structure has always been

Reply:
Good with this then